File: src/main/java/com/fifo/ticketing/domain/user/service/handler/OAuth2LoginSuccessHandler.java
package com.fifo.ticketing.domain.user.service.handler;

import com.fifo.ticketing.domain.user.dto.SessionUser;
import com.fifo.ticketing.domain.user.dto.oauth.UserOAuthDetails;
import com.fifo.ticketing.domain.user.entity.User;
import com.fifo.ticketing.domain.user.repository.UserRepository;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.authentication.AuthenticationSuccessHandler;
import org.springframework.security.web.savedrequest.HttpSessionRequestCache;
import org.springframework.security.web.savedrequest.SavedRequest;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class OAuth2LoginSuccessHandler implements AuthenticationSuccessHandler {

    private final UserRepository userRepository;
    private final HttpSessionRequestCache requestCache = new HttpSessionRequestCache();

    @Override
    public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response,
        Authentication authentication) throws IOException, ServletException {

        UserOAuthDetails userDetails = (UserOAuthDetails) authentication.getPrincipal();

        User user = userRepository.findByEmail(userDetails.getEmail()).orElseThrow();
        request.getSession()
            .setAttribute("loginUser", new SessionUser(user.getId(), user.getUsername()));

        SavedRequest savedRequest = requestCache.getRequest(request, response);
        if (savedRequest != null) {
            String targetUrl = savedRequest.getRedirectUrl();
            response.sendRedirect(targetUrl);
        } else {
            response.sendRedirect("/");
        }
    }
}

File: src/main/java/com/fifo/ticketing/domain/user/service/handler/OAuth2LoginSuccessHandler.kt
package com.fifo.ticketing.domain.user.service.handler

import com.fifo.ticketing.domain.user.dto.SessionUser
import com.fifo.ticketing.domain.user.dto.oauth.UserOAuthDetails
import com.fifo.ticketing.domain.user.entity.Role
import com.fifo.ticketing.domain.user.repository.UserRepository
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import org.springframework.security.web.savedrequest.HttpSessionRequestCache
import org.springframework.stereotype.Component
import java.io.IOException

@Component
class OAuth2LoginSuccessHandler(
    private val userRepository: UserRepository
) : AuthenticationSuccessHandler {

    private val requestCache = HttpSessionRequestCache()

    @Throws(IOException::class, ServletException::class)
    override fun onAuthenticationSuccess(
        request: HttpServletRequest,
        response: HttpServletResponse,
        authentication: Authentication
    ) {
        val userDetails = authentication.principal as UserOAuthDetails

        val user = userRepository.findByEmail(userDetails.email)
        request.session
            .setAttribute(
                "loginUser",
                SessionUser(user!!.id, user.username, user.role)
            )

        val savedRequest = requestCache.getRequest(request, response)
        when {
            savedRequest != null -> response.sendRedirect(savedRequest.redirectUrl)
            user.role == Role.ADMIN -> response.sendRedirect("/admin/menu")
            else -> response.sendRedirect("/performances")
        }
    }
}
